### PR TITLE
PURCHASE PLANNER FINAL PR

### PR DESCRIPTION
**Pull request by marvin-nyalik**
Dear reviewer,

This is my PR for the Ruby on Rails capstone project. Here's a list of what I have done in this project:
- Splash screen

   A simple page with the name of your app (yes, you need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages

   The user should be able to register in the app with full name, email and password (all mandatory).
   The user can log into the app using email and password.
   If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)

   When the user logs in, they are presented with the categories page.
   For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that 
   category.
   When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
   There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page

   For a given category, the list of transactions is presented, ordered by the most recent.
   At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions 
   in that category).
   There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
   When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page

    The user fills out a form to create a new category, indicating their name and icon (both mandatory).
    The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
    When the user clicks on the "Back" button (<), the user navigates to the home page.

-  "Add a new transaction" page

     The user fills out a form to create a new transaction with:
        name (mandatory)
       amount (mandatory)
       categories (mandatory at least one)
       The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that 
       category.
       When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

- I also created unit and integration tests for all the most important components of this application.

Let me know if there are any changes required 👍 @marvin-nyalik

### CODE REVIEW

File: app/views/static_pages/splash.html.erb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<div class="d-flex flex-column justify-content-center align-items-center" style="height: 100vh;">
+  <h1>PURCHASE PLANNER</h1>
+  <p>Please sign in or register to get started:</p>
+  <div>
+    <%= link_to "Sign In", new_user_session_path, class: "btn btn-primary" %>
+    <%= link_to "Register", new_user_registration_path, class: "btn btn-primary" %>
+  </div>
+</div>
```
Comment (zieeco) [inline app/views/static_pages/splash.html.erb]: - Hey, let's improve our design as your design doesn't seem to follow the main design. please, consider making these changes to improve your design:

   - the tow buttons are at the bottom of the container alinged vertically taking the full width with a text content of login and sign up.

*main design*
image


*your design*
image

File: app/views/devise/sessions/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      <div class="form-group">
+        <%= f.submit "Sign in", class: "btn btn-primary" %>
+      </div>
+
+      <div class="form-group">
+        <%= render "devise/shared/links"%>
+      </div>
+    <% end %>
+  </div>
+</div>
```
Comment (zieeco) [inline app/views/devise/sessions/new.html.erb]: - Please, kindly update your design to be close to the main design as seen below:
   - Add the back arrow and maintain the alignment of the contents including the input and how they appear on the design.



*main sign in*
image



*your design*
image

File: app/views/devise/registrations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      <div class="form-group">
+        <%= f.submit "Sign up", class: "btn btn-primary" %>
+      </div>
+
+      <div class="form-group">
+        <%= render "devise/shared/links"%>
+      </div>
+    <% end %>
+  </div>
+</div>
```
Comment (zieeco) [inline app/views/devise/registrations/new.html.erb]: - Let's also improve the sign-up page to follow the design layout, add the missing back arrow, follow the layout design of the inputs to make your design consistent following the main design:


*your design*
image

*main design*
image

File: app/views/categories/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            </div>
+          </div>
+        </div>
+      </div>      
+      <% end %>
+    <% end %>
+  </div>
+  <div class="text-center mt-3">
+    <%= link_to 'Add A New Category', new_category_path, class: "btn btn-primary" %>
+  </div>
```
Comment (zieeco) [inline app/views/categories/index.html.erb]: **FOR ALL BUTTONS  FOR ADDING A TRANSACTION, ADD CATEGORY**


- Hey, kindly move your buttons to the bottom of the container and use the correct background colour as seen in the design.
- Also remove the extra added space on top of the container.

*your design*
image



*main design*
image

File: app/views/categories/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        <%= f.label :icon %>
+        <%= f.text_field :icon, autocomplete: "name", class: "form-control" %>
+      </div>
+    
+      <div class="text-center mt-4">
+        <%= f.submit "Add Category", class: "btn btn-primary" %>
+      </div>
+    </div>
+  <% end %>
+</div>
```
Comment (zieeco) [inline app/views/categories/new.html.erb]: - Kingly remove the space at the top of the container, move the button to the bottom of the container and style the button  according using the correct background, etc.

File: app/views/categories/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            <%= link_to 'Add Transaction', new_category_purchase_path(@category), class: "btn btn-primary" %>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+</div>
+<% else %>
+<h1>You are not authorized to view this page</h1>
+<% end %>
```
Comment (zieeco) [inline app/views/categories/show.html.erb]: - Hey, please, kindly remove the added space at the top of the container and move the button to the bottom of the container. Also, the button should have the correct background as seen from the main design.

File: spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (zieeco) [inline spec/models/user_spec.rb]: - Hi, you've not added any test. please, kindly add them.

File: spec/requests/purchases_spec.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
+
+RSpec.describe 'Purchases', type: :request do
+  describe 'GET /index' do
+    pending "add some examples (or delete) #{__FILE__}"
+  end
+end
```
Comment (zieeco) [inline spec/requests/purchases_spec.rb]: - Hey, also kindly add some tests as you've not added anything yet.

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @marvin-nyalik ,


Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Nice project implmentation.✔️
- Nice design. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._





## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/requests/purchases_spec.rb (hunk 8 above)
Comment (OyePriscilla) [inline spec/requests/purchases_spec.rb]: - [ ] Please as pointed out earlier, kindly ensure that you implement test for this controller (purchases.rb). This code sets up a placeholder for testing the 'GET /index' action in the 'Purchases' controller of a Rails application using RSpec. The actual test examples are yet to be written, and this code serves as a reminder for the developer to add the tests. ℹ️

File: app/views/categories/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,26 @@
+<div class="container mt-5">
+  <div class="d-flex justify-content-between bg-primary text-light align-items-center">
+    <%= link_to categories_path do %>
+      <i class="bx bxs-chevron-left display-4 text-white"></i>
+    <% end %>
+    <h5 class="card-title">Add New Category</h5>
+  </div>
```
Comment (OyePriscilla) [inline app/views/categories/new.html.erb]: - [ ] As raised earlier, kindly remove the top margin around this container. Designing in accordance to template shows how well we pay attention to details. ℹ️ 

Untitled

**Note: This is applicable to `app/views/categories/index.html.erb` and `app/views/categories/show.html.erb` files as pointed out earlier**

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @marvin-nyalik 👏 👏, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close-so-close-seth-meyers

 ## Highlights
Implemented tests for users_model ✔️ 
Designed signup/sign-in in accordance to the given template ✔️ 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- [optional] Kindly take note that you supposed to have opened this PR from feature branch into dev as stated in the project requirement.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ibizugbe — APPROVED** [review]
Hi @marvin-nyalik 👋🏼👋🏼👋🏼, 

approved


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights✅✅✅

- [x] Tests cases for the purchase controller has been captured👍🏼.
- [x] Top margin has been removed from the respective pages pointed by the previous reviewers👍🏼.
- [x] All tests are passing👍🏼.

## Optional suggestions

- [OPTIONAL] Kindly take note that you are meant to make use of gitflow and not github flow for this project. reasons been that it is a solo capstone😉. Try to always look at the flow requirements before opening a PR.
image


_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**